src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -398,20 +399,40 @@ private int checkArraySize(String typeName, int elemSize) {
     }
 
     @ForceInline
-    void checkBounds(long offset, long length) {
+    void checkSliceBounds(long offset, long length) {
+        try {
+            checkBounds(offset, length);
+        } catch (IndexOutOfBoundsException e) {
+            throwOutOfBounds(offset, length, /* isSlice = */ true);
+        }
+    }
+
+    @ForceInline
+    void checkAccessBounds(long offset, long length) {
+        try {
+            checkBounds(offset, length);
+        } catch (IndexOutOfBoundsException e) {
+            throwOutOfBounds(offset, length, /* isSlice = */ false);
+        }
+    }
+
+    @ForceInline
+    private void checkBounds(long offset, long length) {
         if (length > 0) {
```

Comment:
Note: if we know we're not coming from a slice operation, we can just drop the if... so I wonder if we could split the impl in two to remove extra branches in the access case?